internal/runner/accept.go

```diff
@@ -89,63 +88,131 @@ func (r *Runner) AcceptExact(ctx context.Context, id string) error {
 	if tok := serverclient.TokenFromContext(r.runningCtx); tok != "" {
 		ctx = serverclient.TokenWithContext(ctx, tok)
 	}
 
-	// Open a new job stream. This retries on connection errors. Note that
-	// this loop doesn't respect the accept timeout because gRPC has no way
-	// to time out of a "WaitForReady" RPC call (it ignores context cancellation,
-	// too). TODO: do a manual backoff with WaitForReady(false) so we can
-	// weave in accept timeout.
+	// Retry tracks whether we're trying a job stream connection or not.
+	// We use this so that the first attempt fails fast so we can log it.
+	// Subsequent attempts block.
 	retry := false
+
+	// State we want to initialize outside the retry label.
 	var client pb.Waypoint_RunnerJobStreamClient
-	for {
-		// Get our configuration state value. We use this so that we can detect
-		// when we've reconnected during failures.
-		stateGen := r.readState(&r.stateConfig)
-
-		// NOTE: we purposely do NOT use ctx above since if the context is
-		// cancelled we want to continue reporting errors.
-		log.Debug("opening job stream", "retry", retry)
-		var err error
-		client, err = r.client.RunnerJobStream(streamCtx, grpc.WaitForReady(retry))
-		if err != nil {
-			if status.Code(err) == codes.Unavailable || status.Code(err) == codes.NotFound {
-				log.Warn("server down during job stream open, will attempt reconnect")
-
-				// Since this is a disconnect, we have to wait for our
-				// RunnerConfig stream to re-establish. We wait for the config
-				// generation to increment.
-				if r.waitStateGreater(&r.stateConfig, stateGen) {
-					return status.Error(codes.Internal, "early exit while waiting for reconnect")
-				}
+	var streamCtx context.Context
+	var streamCancel context.CancelFunc
+	var streamCtxLock sync.Mutex
+	var stateGen uint64
+	var err error
+
+	// We wrap this in a func() so that we use the latest client value
+	// and don't stack defers on retry.
+	defer func() {
+		if client != nil {
+			client.CloseSend()
+		}
 
-				retry = true
-				continue
-			}
+		streamCtxLock.Lock()
+		defer streamCtxLock.Unlock()
+		if streamCancel != nil {
+			// Wrap in a func() so that if we retry, we don't stack defers.
+			streamCancel()
+		}
+	}()
 
-			return err
+	// If we have a timeout, then we setup a timer for accepting.
+	var acceptTimer *time.Timer
+	var canceled int32
+	if r.acceptTimeout > 0 {
+		acceptTimer = time.AfterFunc(r.acceptTimeout, func() {
+			log.Error("runner timed out waiting for a job",
+				"timeout", r.acceptTimeout.String())
+
+			// Grab this lock before updating canceled. You don't
+			// need to have this lock to touch canceled (we use atomic ops)
+			// but it is used when the streamCancel is being reset so that
+			// we don't set it up and race with cancelation.
+			streamCtxLock.Lock()
+			defer streamCtxLock.Unlock()
+
+			// Mark that we canceled
+			atomic.StoreInt32(&canceled, 1)
+
+			// Cancel the context
+			streamCancel()
```

Comment:
Do we need to check that `streamCancel` is not nil? Most of the other places do...

Reply:
Yep, good find.